src/main/noise/MLS.cpp

```diff
@@ -0,0 +1,318 @@
+                8193u,
+                402653187u,
+                12291u,
+                8193u,
+                7083549724304467820547u,
+                274877906945u,
+                786435u,
+                29014219670751100192948227u,
+                12291u,
+                5u,
+                2097153u,
+                2049u,
+                703687441776645u,
+                65u,
+                2049u,
+                175921860444165u,
+                137438953473u,
+                195u,
+                226673591177742970257411u,
+                513u,
+                3075u,
+                65537u,
+                32769u,
+                46116860184273879045u,
+                2147483649u,
+                195u,
+                12291u,
+                1025u,
+                43980465111045u,
+                513u,
+                7253554917687775048237059u,
+                49155u,
+                3541774862152233910275u,
+                1310725u,
+                8589934593u,
+                257u,
+                334903147375496382040217013234696321u,
+                262145u,
+                1729382256910270467u,
+                5u,
+                137438953473u,
+                486777830487640090174734030864387u,
+                206158430211u,
+                3u,
+                671088645u
+                #endif
+        };
+
+        MLS::MLS()
+        {
+            construct();
+        }
+
+        MLS::~MLS()
+        {
+            destroy();
+        }
+
+        void MLS::construct()
+        {
+            nBits               = sizeof(mls_t) * 8;
+            nFeedbackBit        = 0;
+            nFeedbackMask       = 0;
+            nActiveMask         = 0;
+            nTapsMask           = 0;
+            nOutputMask         = 1;
+            nState              = 0;
+
+            fAmplitude          = 1.0f;
+
+            bSync               = true;
+        }
+
+        void MLS::destroy()
+        {
+        }
+
+        size_t MLS::maximum_number_of_bits()
+        {
+            return nMaxBits;
+        }
+
+        void MLS::update_settings()
+        {
+            if (!needs_update())
+                return;
+
+            nBits = lsp_max(nBits, 1u);
+            nBits = lsp_min(nBits, nMaxBits);
+
+            nFeedbackBit = nBits - 1u;
+            nFeedbackMask = mls_t(1) << nFeedbackBit;
+
+            // Switch on all the first nBits bits.
+            nActiveMask = ~(~mls_t(0) << nBits);
+
+            nTapsMask = vTapsMaskTable[nBits - 1];
+
+            nState &= nActiveMask;
+
+            // State cannot be 0. If that happens, flip all the active bits to 1.
+            if (nState == 0)
+                nState |= nActiveMask;
+
+            bSync = false;
+        }
+
+        // Compute the xor of all the bits in value
+        MLS::mls_t MLS::xor_gate(mls_t value)
+        {
+            mls_t nXorValue;
+            for (nXorValue = 0; value; nXorValue ^= 1)
```

Comment:
This algorithm has linear complexity for the worst case (when all bits are ones).
I think it can be replaced with solution of logarithmic complexity:

```c++
/* Note: value should be unsigned for proper binary shift calculations */
#ifdef ARCH_128_BIT
value = (value >> 64) ^ (value & 0xffffffffffffffff);
#endif /* ARCH_128_BIT */
#ifdef ARCH_64_BIT
value = (value >> 32) ^ (value & 0xffffffff);
#endif /* ARCH_64_BIT */
value = (value >> 16) ^ (value & 0xffff);
// The code below can be replaced with lookup table of 256 elements
value = (value >> 8) ^ (value & 0xff);
value = (value >> 4) ^ (value & 0xf);
value = (value >> 2) ^ (value & 0x3);
value = (value >> 1) ^ (value & 0x1);
```